dsp/modules/gpt3.py

```diff
@@ -20,21 +21,43 @@ def backoff_hdlr(details):
 
 
 class GPT3(LM):
-    """Wrapper around OpenAI's GPT-3 API.
-
-    Currently supported models include `davinci`, `curie`, `babbage`, `ada`, `gpt-3.5-turbo`, and `gpt-3.5-turbo-0301`.
+    """Wrapper around OpenAI's GPT API. Supports both the OpenAI and Azure APIs.
+
+    Args:
+        model (str, optional): OpenAI or Azure supported LLM model to use. Defaults to "text-davinci-002".
+        api_key (Optional[str], optional): API provider Authentication token. use Defaults to None.
+        api_provider (Literal["openai", "azure"], optional): The API provider to use. Defaults to "openai".
+        model_type (Literal["chat", "text"], optional): The type of model that was specified. Mainly to decide the optimal prompting strategy. Defaults to "text".
+        **kwargs: Additional arguments to pass to the API provider.
     """
 
     def __init__(
-        self, model: str = "text-davinci-002", api_key: Optional[str] = None, **kwargs
+        self,
+        model: str = "text-davinci-002",
+        api_key: Optional[str] = None,
+        api_provider: Literal["openai", "azure"] = "openai",
+        model_type: Literal["chat", "text"] = "text",
+        **kwargs,
     ):
         super().__init__(model)
         self.provider = "openai"
+        self.model_type = model_type
+
+        if api_provider == "azure":
+            assert (
+                "engine" in kwargs or "deployment_id" in kwargs
+            ), "Must specify engine or deployment_id for Azure API instead of model."
+            assert "api_version" in kwargs, "Must specify api_version for Azure API"
+            assert "api_base" in kwargs, "Must specify api_base for Azure API"
+            openai.api_type = "azure"
+            openai.api_base = kwargs["api_base"]
+            if kwargs.get("api_version"):
```

Comment:
This might be a bit redundant with the assert above. Not sure what's better to drop (assert or if).

Reply:
ah, yes. I missed removing this. Initially, I was under the impression that api_version was optional, so added this if case. Will remove this in a further PR.